### ISSUE TITLE
wrap() does not refocus wrapped element

### ISSUE THREAD

**Issue by idontusenumbers** [jquery/jquery#3165]
After calling wrap() on a textbox that has focus, focus is not reassigned to that element after wrapping is complete and instead the element loses focus.

Test case; textbox should stay green:
https://jsfiddle.net/915jwh26/

Here's a workaround/possible solution:

```
    var inputWasFocused = $input.is(':focus');
    $input.wrap($('<span>');
    if (inputWasFocused)
        $input.focus();
```

**Comment by hehexianshi** [comment]
I think if input is focus used after wrap should also be the state of focus, sorry my English is not good
like this：
`$(function(){
	$('input').click(function() {
		$('input').wrap('<span>')  
  })
})`

**Comment by dmethvin** [comment]
This problem generally applies to any situation where we remove content from the document to manipulate it, I'm not sure if wrap is the only place where that happens. That test case is the easy one where the thing being wrapped has focus, in general the focus could be on an element inside the wrapped structure, like this: https://jsfiddle.net/915jwh26/1/ For that case we'd need to see if the currently focused element is inside the content to be wrapped then refocus it after wrapping. This can be risky in some cases, see `safeActiveElement` which is there for IE9.

**Comment by timmywil** [comment]
I think the workaround is the best way to handle this. However, we'd be willing to review a patch.

**Comment by idontusenumbers** [comment]
Why is this being closed without a resolution =(

**Comment by timmywil** [comment]
The resolution is patchwelcome. It is low priority, probably not worth adding code to jQuery, and has a suitable workaround. There are probably better solutions that don't require wrapping an element that currently has focus.

**Comment by idontusenumbers** [comment]
If it's closed, it will never get fixed because no one will know the problem exists. What reason is there to close it? Is there some pressure to resolve open tickets?

As it stands, the description of the wrap method does not include a mention that the element will get removed from the DOM and thus trigger strange unexpected side effects like this. Placing a note on the bottom of the doc with the workaround is just going to frustrate developers.

**Comment by dmethvin** [comment]
Well, we did say a patch is welcome, and the ticket can be found through search whether it's open or not. We only tend to keep tickets open that we think we could solve. Since the issue has been around for a decade with no previous reports that I can recall it's not very high priority.

There are several issues here that should be addressed by a patch.

1) As mentioned above, it is likely the problem exists elsewhere when using other APIs. For example, any use of `.detach()` will remove focus from an element, even if the elements removed are later re-inserted into the document. How should that be treated, either docs or code? What other cases exist in the API? In general when we don't treat cases consistently it leads to further bug reports.

2) Focus management is extremely fidgety in several browsers, it's so bad in Firefox that we had to [disable several flakey tests](https://github.com/jquery/jquery/blob/25d8ccd1112d75394b91071ff7eba13283aaf898/test/unit/event.js#L2985) and we [can't synchronously focus elements in IE10](https://github.com/jquery/jquery/blob/25d8ccd1112d75394b91071ff7eba13283aaf898/test/unit/event.js#L3024).. With those obstacles, any time jQuery tries to force focus somewhere it's quite possible we will screw it up. In contrast,  you know for sure where you want focus to be and it's more straightforward to do it at the application level.

**Comment by idontusenumbers** [comment]
I can see your point about the detach method. I think the wrap method is a little different in that it should atomically remove and then re-insert into the DOM.

I also understand the concerns about focus handling fidgetyness.

I'm not familiar with the jQuery code base, so I feel apprehensive about submitting a patch. If the workaround has no side effects, is there any concern with just integrating it into the code base?

**Comment by dmethvin** [comment]
> I think the wrap method is a little different in that it should atomically remove and then re-insert into the DOM.

I haven't thought about other APIs that might have the same situation, at minimum [all the wrap and unwrap methods](https://github.com/jquery/jquery/blob/25d8ccd1112d75394b91071ff7eba13283aaf898/src/wrap.js) are affected.

> If the workaround has no side effects, is there any concern with just integrating it into the code base?

My spidey sense says that there will be side effects. But if the patch was small and worked on all supported browsers we'd certainly consider it. When you think about how many times a day jQuery is loaded (trillions?) it would be hard to justify adding for example 100 bytes to fix a problem that one person has reported and can be worked around easily.
